File: pynet/pynet2_mult.py
#!/usr/bin/env python3

# This script splits a large bulk of xmls
# for processing it in parallel

import os
import sys
import signal
import argparse
import platform
import subprocess
import multiprocessing
from multiprocessing import Process, Pool

# Global list
delay = 5
processes = []
splitted_files = []
number_processes = 8    # default value
#pynet2_executable = os.getcwd() + "/pynet2_obf.py"
#pynet2_executable = os.path.join(os.getcwd(), "pynet2_obf.py")
#pynet2_executable = os.path.join(os.getcwd(), "pynet2.py")
pynet2_executable = os.path.abspath("pynet2.py")


def exit_gracefully():
    print("\n\n[!] Exiting gracefully...")

    # Terminate all processes
    for p in processes:
        p.terminate()
    for p in processes:
        p.join()

    # Clear the global list of processes
    processes.clear()

    sys.exit(0)


# Signal handler function to catch Ctrl+C
def signal_handler(sig, frame):
    exit_gracefully()


# Register the signal handler for SIGINT (Ctrl+C)
signal.signal(signal.SIGINT, signal_handler)


# Function to parse command-line arguments
def parse_arguments():
    parser = argparse.ArgumentParser(description="PyNet - A Python-based tool for telnet and network operations.")
    parser.add_argument("-iL", "--ip_list", help="\t\tPath to a file containing a list of IP addresses or hostnames, one per line.")
    parser.add_argument("-iC", "--commands_list", help="\t\tPath to a file containing a list of telnet commands to execute, one per line.")
    parser.add_argument("-n", "--processes", help="\t\tNumber of processes.")
    parser.add_argument("-i", "--ip", help="\t\tTarget IP address or hostname to authenticate with.")
    parser.add_argument("-p", "--port", default=23, type=int, help="\t\tTarget port (default is 23 for Telnet).")
    parser.add_argument("-l", "--login", action="store_true", help="\tSimply login on the target system.")
    parser.add_argument("-v", "--victim", help="\tTarget victim for sending attacks (IP or domain name).")
    parser.add_argument("-c", "--command", help="\tCommand to execute on the target system.")
    parser.add_argument("-d", "--delay", help="\tDelay in seconds between listed hosts for performing connections.")

    return parser.parse_args()


# Method for splitting a file
def split_file(input_path, number):

    with open(input_path) as file:
        ip_addresses = file.read().splitlines()

        total_lines = len(ip_addresses)
        chunk_size = int(total_lines / number)

        cont = 0
        cont_less = 1
        buffer = list()
        for ip in ip_addresses:

            # Refresh variables when chunk size is encountered
            if cont == chunk_size:

                out_path = input_path + f".{cont_less}"
                splitted_files.append(out_path)
                print(f"[!] Splitted: {out_path}     {len(buffer)}")

                #write the buffer into a file
                with open(out_path, 'w') as buffer_file:
                    buffered_string = '\n'.join(buffer)
                    buffer_file.write(buffered_string)

                # Update variables
                cont_less += 1
                cont = 0
                buffer = []
            else:
                buffer.append(ip)
                cont += 1
                #print(ip)

        # Last buffered when not reach chunk_size
        out_path = input_path + f".{cont_less}"
        splitted_files.append(out_path)
        print(f"[!] Splitted: {out_path}     {len(buffer)}")

        #write the buffer into a file
        with open(out_path, 'w') as buffer_file:
            buffered_string = '\n'.join(buffer)
            buffer_file.write(buffered_string)

    file.close()



# Call a single pynet
#def call_pynet(ip_file, port, victim, sequence_path):
    #global pynet2_executable
    #os.system(f"python {pynet2_executable} -iL {ip_file} -p {port} -l -v {victim}")
    #os.system(f"python {pynet2_executable} -iL {ip_file} -p {port} -l -v {victim} -iC {sequence_path}")


def call_pynet(ip_file, port, victim, sequence_path):
    global pynet2_executable
    global delay


    command = [
        #"python", pynet2_executable, "-iL", ip_file, "-p", str(port), "-l", "-iC", sequence_path, "--delay 3"
        #"python", pynet2_executable, "-iL", ip_file, "-p", str(port), "-l", "-iC", sequence_path
        sys.executable, pynet2_executable, "-iL", ip_file, "-p", str(port), "-l", "-iC", sequence_path
    ]

    print(f"Command: {command}")  # Before calling subprocess.run
    if None in command:
        print(f"Error: command contains None values: {command}")
        return
    
    print(f"[+] Launching: {' '.join(command)}")
    result = subprocess.run(command, capture_output=True, text=True)
    
    if result.returncode != 0:
        print(f"[!] Error: {result.stderr}")
    else:
        print(f"[✓] Done: {ip_file}")


# Call the multiprocessing method for parallelizing
if __name__ == "__main__":
    # Parse arguments once, only in parent process
    args = parse_arguments()

    # Store arguments somewhere safe
    from functools import partial
    import multiprocessing

    # Split files
    if args.processes:
        number_processes = int(args.processes)

    if args.delay:
        delay = int(args.delay)

    if not os.path.exists(args.ip_list):
        print(f"[!] Path doesn't exist:  {args.ip_list}")
        sys.exit(0)

    if not os.path.exists(args.commands_list):
        print(f"[!] Path doesn't exist:  {args.commands_list}")
        sys.exit(0)

    split_file(args.ip_list, number_processes)

    # Determine the correct context based on the OS
    if platform.system() == "Windows":
        mp_context = multiprocessing.get_context("spawn")
    else:
        mp_context = multiprocessing.get_context("fork")

    # Setup multiprocessing pool safely
    try:
        with mp_context.Pool(number_processes) as pool:
            pool.starmap(
                call_pynet,
                [(ip_file, args.port, args.victim, args.commands_list) for ip_file in splitted_files]
            )
    except KeyboardInterrupt:
        print("[!] Interrupted by user.")
        exit_gracefully()

    #for ip_file in splitted_files:
    #    p = Process(target=call_pynet, args=(ip_file, args.port, args.victim, args.commands_list))
    #    processes.append(p)
    #    p.start()

    #print(Colors.GREEN + f"\n[!] Command executes on {ip_file} begins with {number_processes} processes" + Colors.R)

    # Wait for all processes to complete
    #for p in processes:
    #    p.join()

File: pynet/pynet2_mult.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <time.h>

#define MAX_LINE_LENGTH 1024
#define MAX_SPLITS 10000

char *splitted_files[MAX_SPLITS];
int split_count = 0;
pid_t child_pids[MAX_SPLITS];
int child_count = 0;

// Function to handle SIGINT (Ctrl+C)
void handle_sigint(int sig) {
    printf("\n[!] Caught SIGINT. Terminating child processes...\n");
    for (int i = 0; i < child_count; i++) {
        kill(child_pids[i], SIGTERM);
    }
    exit(1);
}

// Function to split the input file into multiple files
void split_file(const char *input_path, int number) {
    FILE *file = fopen(input_path, "r");
    if (!file) {
        perror("fopen");
        exit(1);
    }

    char **ip_addresses = NULL;
    int total_lines = 0;
    char line[MAX_LINE_LENGTH];

    while (fgets(line, sizeof(line), file)) {
        ip_addresses = realloc(ip_addresses, sizeof(char*) * (total_lines + 1));
        ip_addresses[total_lines] = strdup(line);
        total_lines++;
    }
    fclose(file);

    // Seed random generator
    srand(time(NULL));

    // Fisher-Yates shuffle
    for (int i = total_lines - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        char *temp = ip_addresses[i];
        ip_addresses[i] = ip_addresses[j];
        ip_addresses[j] = temp;
    }

    int chunk_size = total_lines / number;
    int remainder = total_lines % number;
    int index = 0;

    for (int i = 0; i < number; i++) {
        int current_chunk_size = chunk_size + (i < remainder ? 1 : 0);
        char *out_path = malloc(strlen(input_path) + 15);
        sprintf(out_path, "%s.%d", input_path, i + 1);
        FILE *out_file = fopen(out_path, "w");
        if (!out_file) {
            perror("fopen");
            exit(1);
        }
        for (int j = 0; j < current_chunk_size; j++) {
            fputs(ip_addresses[index], out_file);
            free(ip_addresses[index]);
            index++;
        }
        fclose(out_file);
        splitted_files[split_count++] = out_path;
        printf("[!] Splitted: %s     %d\n", out_path, current_chunk_size);
    }
    free(ip_addresses);
}

// Function to call pynet2.py with appropriate arguments
void call_pynet(const char *ip_file, const char *port, const char *commands_list, const char *g_value, const char *d_value) {
    pid_t pid = fork();
    if (pid == 0) {
        // Child process

        // Print the command to be executed
        printf("Executing: run_pynet2.sh -iL %s -p %s -l -iC %s -g %s -d %s\n", ip_file, port, commands_list, g_value, d_value);
        fflush(stdout);

        // Redirect stdout and stderr to /dev/null
        int devnull = open("/dev/null", O_WRONLY);
        if (devnull != -1) {
            dup2(devnull, STDOUT_FILENO);
            dup2(devnull, STDERR_FILENO);
            close(devnull);
        }

        execlp("/media/Kali/home/grimaldi/Bash/Telnet/pynet/run_pynet2.sh",
               "run_pynet2.sh", "-iL", ip_file, "-p", port, "-l", "-iC", commands_list, "-g", g_value, "-d", d_value,
               (char *)NULL);

        // Only reached if execlp fails
        perror("execlp");
        exit(1);
    } else if (pid > 0) {
        // Parent process
        child_pids[child_count++] = pid;
    } else {
        // Fork failed
        perror("fork");
        exit(1);
    }
}

int main(int argc, char *argv[]) {
    if (argc < 13) {
        fprintf(stderr, "Usage: %s -iL <ip_list> -iC <commands_list> -n <processes> -g <group_size> -d <depth> [-p <port>]\n", argv[0]);
        exit(1);
    }

    char *ip_list = NULL;
    char *commands_list = NULL;
    char *processes = NULL;
    char *port = "23"; // Default port
    char *g_value = NULL;
    char *d_value = NULL;

    // Parse command-line arguments
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-iL") == 0 && i + 1 < argc) {
            ip_list = argv[++i];
        } else if (strcmp(argv[i], "-iC") == 0 && i + 1 < argc) {
            commands_list = argv[++i];
        } else if (strcmp(argv[i], "-n") == 0 && i + 1 < argc) {
            processes = argv[++i];
        } else if (strcmp(argv[i], "-g") == 0 && i + 1 < argc) {
            g_value = argv[++i];
        } else if (strcmp(argv[i], "-d") == 0 && i + 1 < argc) {
            d_value = argv[++i];
        } else if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) {
            port = argv[++i];
        }
    }

    if (!ip_list || !commands_list || !processes || !g_value || !d_value) {
        fprintf(stderr, "Missing required arguments.\n");
        exit(1);
    }

    int num_processes = atoi(processes);
    if (num_processes <= 0 || num_processes > MAX_SPLITS) {
        fprintf(stderr, "Invalid number of processes. Must be between 1 and %d.\n", MAX_SPLITS);
        exit(1);
    }

    // Register SIGINT handler
    signal(SIGINT, handle_sigint);

    // Split the input file
    split_file(ip_list, num_processes);

    // Fork child processes to call pynet2.py
    for (int i = 0; i < split_count; i++) {
        call_pynet(splitted_files[i], port, commands_list, g_value, d_value);
    }

    // Wait for all child processes to finish
    for (int i = 0; i < child_count; i++) {
        waitpid(child_pids[i], NULL, 0);
    }

    // Free allocated memory
    for (int i = 0; i < split_count; i++) {
        free(splitted_files[i]);
    }

    return 0;
}
